src/storage/slice.hpp

```diff
@@ -10,27 +10,27 @@
 
 namespace storage
 {
-    /// Create a new storage object by slicing the given storage object.
-    ///
-    /// Example of slicing a storage with offset 3 and size 2:
-    ///
-    /// Given storage:     [0, 1, 2, 3, 4, 5]
-    /// Returned storage:  [3, 4]
-    ///
-    /// This function makes sure the bounds of the initial storage object is not
-    /// exceeded.
-    ///
-    /// @param storage The storage object to be used for the slice operation
-    /// @param offset The number of bytes to offset the given storage object,
-    ///               must be smaller than the storage object's size.
-    /// @param size The number of bytes to include in the slice,
-    ///               must be smaller than the offset storage object's size.
-    /// @return A new, sliced storage object.
-    template<class Storage>
-    Storage slice(const Storage& storage, uint32_t offset, uint32_t size)
-    {
-        assert(offset <= storage.size());
-        assert(size <= storage.size() - offset);
-        return Storage(storage.data() + offset, size);
-    }
+/// Create a new storage object by slicing the given storage object.
+///
+/// Example of slicing a storage with offset 3 and size 2:
+///
+/// Given storage:     [0, 1, 2, 3, 4, 5]
+/// Returned storage:  [3, 4]
+///
+/// This function makes sure the bounds of the initial storage object is not
+/// exceeded.
+///
+/// @param storage The storage object to be used for the slice operation
+/// @param offset The number of bytes to offset the given storage object,
+///               must be smaller than the storage object's size.
+/// @param size The number of bytes to include in the slice,
+///               must be smaller than the offset storage object's size.
+/// @return A new, sliced storage object.
+template<class Storage>
+Storage slice(const Storage& storage, uint32_t offset, uint32_t size)
+{
+    assert(offset <= storage.size());
```

Comment:
I think the asserts are wrong here. Offset must be smaller than size to avoid getting a zero size storage object. I guess this can be implemented in terms of `offset` and `shrink`?

```
auto tmp = ::offset(storage, offset);
return shrink(tmp, size);
```